packages/typescript-estree/src/tsconfig-parser.ts

```diff
@@ -146,19 +198,80 @@ export function calculateProjectParserOptions(
       }
     };
 
-    // register callbacks to trigger program updates without using fileWatchers
-    watchCompilerHost.watchFile = (fileName, callback): ts.FileWatcher => {
+    // in watch mode, eslint will give us the latest file contents
+    // store the watch callback so we can trigger an update with eslint's content
+    watchCompilerHost.watchFile = (
+      fileName,
+      callback,
+      interval,
+    ): ts.FileWatcher => {
+      // specifically (and separately) watch the tsconfig file
+      // this allows us to react to changes in the tsconfig's include/exclude options
+      let watcher: Watcher | null = null;
+      if (fileName.includes(tsconfigPath)) {
+        watcher = watch(
+          fileName,
+          {
+            interval,
+          },
+          extra,
+        );
+        watcher.on('change', path => {
+          callback(path, ts.FileWatcherEventKind.Changed);
+        });
+        configSystemFileWatcherTrackingSet.add(watcher);
+      }
+
       const normalizedFileName = path.normalize(fileName);
-      watchCallbackTrackingMap.set(normalizedFileName, callback);
+      const watchers = ((): Set<ts.FileWatcherCallback> => {
+        let watchers = watchCallbackTrackingMap.get(normalizedFileName);
+        if (!watchers) {
+          watchers = new Set();
+          watchCallbackTrackingMap.set(normalizedFileName, watchers);
+        }
+        return watchers;
+      })();
+      watchers.add(callback);
+
       return {
         close: (): void => {
-          watchCallbackTrackingMap.delete(normalizedFileName);
+          watchers.delete(callback);
+
+          if (watcher) {
+            watcher.close();
+            configSystemFileWatcherTrackingSet.delete(watcher);
+          }
         },
       };
     };
 
-    // ensure fileWatchers aren't created for directories
-    watchCompilerHost.watchDirectory = (): ts.FileWatcher => noopFileWatcher;
+    // when new files are added in watch mode, we need to tell typescript about those files
+    // if we don't then typescript will act like they don't exist.
+    watchCompilerHost.watchDirectory = (
```

Comment:
funny story here - typescript aggressively finds, and resolves symlinks in `node_modules`.
i.e. for `eslint-plugin`, it will ask us to watch these paths:
- packages/eslint-plugin/src
- packages/eslint-plugin/tests
- packages/eslint-plugin/node_modules
- packages/eslint-plugin/node_modules/@types
- packages/experimental-utils
- packages/parser
- packages/typescript-estree

The final 3 occur because lerna creates symlinks because they are monorepo deps. Because they are symlinks typescript specifically asks us to watch them, where as the other folders in `node_modules` are considered handled by the first item in the list.